### PR TITLE
required and number validations

### PR DESCRIPTION
**Pull request by 0xF013**
https://community.openproject.org/work_packages/19713
https://community.openproject.org/work_packages/19712
https://community.openproject.org/work_packages/19630

required and number validation
please test german version
also check out dem fancy starts

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://0xf013-feature-inplace-validations.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://0xf013-feature-inplace-validations.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by NobodysNightmare** [comment]
You should probably only display the star for fields that are not only required, but also writable.

For me the "spent time" is currently displayed as required, though you will never be able to set it.

**Comment by 0xF013** [comment]
@NobodysNightmare will fix it (why is it even required: true then?)

**Comment by NobodysNightmare** [comment]
`required` is defined as "can't be null" (API-wise). And indeed the spent time can't be null...

**Comment by 0xF013** [comment]
@NobodysNightmare please set it to FileNotFound http://thedailywtf.com/articles/What_Is_Truth_0x3f_

**Comment by 0xF013** [comment]
@NobodysNightmare fixed

**Comment by NobodysNightmare** [comment]
This is the time when I reference actual documentation from the MSDN: https://msdn.microsoft.com/de-de/library/microsoft.office.core.msotristate%28v=office.11%29.aspx

**Comment by myabc** [comment]
@NobodysNightmare please don't.

File: frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js [hunk 1]
```diff
@@ -40,8 +40,27 @@ module.exports = function(
...
+        if (vm.editForm.$invalid) {
+          var detectedViolations = [];
+          acknowledgedValidationErrors.forEach(function(error) {
+            if (vm.editForm.$error[error]) {
+              detectedViolations.push(I18n.t('js.inplace.errors.' + error, {
+                field: fieldController.getLabel()
+              }));
+            }
+          });
+          if (detectedViolations.length) {
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js:57]: Urgh... do they still do it like C in JavaScript?

I would have written `> 0`, but I guess this is optional. Do we have a styleguide for such things?

edit: That means, I am not sure if this is supposed to be beautiful...

**Comment by NobodysNightmare** [comment]
If this C-Style if statement is valid, this LGTM (if travis agrees)...

File: frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js (hunk 1 above)
Comment (0xF013) [inline frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js:57]: @NobodysNightmare this stuff here is pretty mild. You might wanna check out NaN detection (x != x)

**Comment by floriank** [comment]
took the liberty and restarted the failing job

File: frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js (hunk 1 above)
Comment (floriank) [inline frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js:57]: Casting any number to boolean is `true`, except for `0` and `NaN`, so this is okay

``` javascript
console.log(!!0); // false
console.log(!!NaN); // false
console.log(!!42); // true
```

And yes, `NaN` is a number:

``` javascript
console.log(typeof NaN); // "number"
```

**Comment by NobodysNightmare** [comment]
I restarted the MySQL specs multiple times now, but they always seem to fail after the git operation (last log message is "Counting objects")...

Any idea? Can't imagine, that this is the fault of this PR, but also weird...

**Comment by NobodysNightmare** [comment]
@0xF013  Now you need to rebase it... I hope that fixes travis ^^

**Comment by 0xF013** [comment]
@NobodysNightmare done, but please click around a bit, there were a couple of coflicts

File: frontend/app/templates/components/inplace_editor/editable/textarea.html [hunk 2]
```diff
@@ -2,7 +2,8 @@
           class="focus-input inplace-edit--textarea"
           name="value"
           ng-disabled="fieldController.state.isBusy"
+          ng-required="editPaneController.isRequired"
           ng-model="fieldController.writeValue.raw"
-          title="{{ editTitle }}"
+          title="{{ fieldController.editTitle }}"
```
Comment (NobodysNightmare) [inline frontend/app/templates/components/inplace_editor/editable/textarea.html:7]: You were talking about merge-conflicts: Not sure if this is one, that got resolved badly...

Could also be an intended change, I am unsure...

Comment (0xF013) [inline frontend/app/templates/components/inplace_editor/editable/textarea.html:7]: that's a quick fix

**Comment by NobodysNightmare** [comment]
I just discovered that the title of all edit fields is wrong ^^

e.g. it is not "% done: Edit", but "percentageDone: Edit" or "customField11: Edit"...

I will probably create a new Bug ticket for that...

**Comment by NobodysNightmare** [comment]
@0xF013 Looking good so far. Waiting for travis and giving you time to answer whether you think the change in textarea.html was intended, I am not sure what it does...

**Comment by floriank** [comment]
Restarted the build

**Comment by NobodysNightmare** [comment]
Restarted karma once more. If it is red, it probably means red.

**Comment by NobodysNightmare** [comment]
@0xF013 it is red when there is a cross instead of a checkmark... I need to make sure my comments are accessible.

**Comment by 0xF013** [comment]
@NobodysNightmare I guess it's ok, I see the red checkmark

**Pull request merged**